File: blue_carbon_app/lib/models/project.dart
class Project {
  final String projectId;
  final String name;
  final String description;
  final String location;
  final String type;
  final String imageUrl;
  final String ngoName;
  final String verificationStatus;
  final double carbonCredits;
  final double confidenceScore;
  final DateTime createdAt;
  final DateTime? verifiedAt;

  const Project({
    required this.projectId,
    required this.name,
    required this.description,
    required this.location,
    required this.type,
    required this.imageUrl,
    required this.ngoName,
    required this.verificationStatus,
    required this.carbonCredits,
    required this.confidenceScore,
    required this.createdAt,
    this.verifiedAt,
  });

  factory Project.fromJson(Map<String, dynamic> json) {
    return Project(
      projectId: json['projectId'] ?? '',
      name: json['projectName'] ?? '',
      description: json['projectDescription'] ?? '',
      location: json['projectLocation'] ?? '',
      type: json['projectType'] ?? '',
      imageUrl: json['imageUrl'] ?? 'assets/images/default_project.png',
      ngoName: json['ngoId']?['organizationName'] ?? 'Unknown NGO',
      verificationStatus: json['verificationStatus'] ?? 'pending',
      carbonCredits: (json['mlAnalysis']?['carbonCreditsCalculated'] ?? 0).toDouble(),
      confidenceScore: (json['mlAnalysis']?['confidenceScore'] ?? 0).toDouble(),
      createdAt: DateTime.parse(json['createdAt'] ?? DateTime.now().toIso8601String()),
      verifiedAt: json['verificationDetails']?['verifiedAt'] != null 
          ? DateTime.parse(json['verificationDetails']['verifiedAt'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'projectId': projectId,
      'projectName': name,
      'projectDescription': description,
      'projectLocation': location,
      'projectType': type,
      'imageUrl': imageUrl,
      'ngoName': ngoName,
      'verificationStatus': verificationStatus,
      'carbonCredits': carbonCredits,
      'confidenceScore': confidenceScore,
      'createdAt': createdAt.toIso8601String(),
      'verifiedAt': verifiedAt?.toIso8601String(),
    };
  }

  String get statusColor {
    switch (verificationStatus) {
      case 'approved':
        return '#28a745';
      case 'rejected':
        return '#dc3545';
      case 'under_review':
        return '#17a2b8';
      default:
        return '#ffc107';
    }
  }

  String get formattedStatus {
    return verificationStatus.replaceAll('_', ' ').toUpperCase();
  }
}

File: blockchain-backend/models/Project.js
const mongoose = require('mongoose');

const projectSchema = new mongoose.Schema({
  // Project identification
  projectId: {
    type: String,
    required: true,
    unique: true,
    index: true
  },
  ngoId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'NGO',
    required: true
  },
  
  // Project details
  projectName: {
    type: String,
    required: true
  },
  projectDescription: {
    type: String,
    required: true
  },
  projectLocation: {
    type: String,
    required: true
  },
  projectType: {
    type: String,
    enum: ['mangrove', 'seagrass', 'saltmarsh', 'other'],
    required: true
  },
  
  // Media files
  beforeImages: [{
    filename: String,
    originalName: String,
    path: String,
    uploadedAt: Date
  }],
  afterImages: [{
    filename: String,
    originalName: String,
    path: String,
    uploadedAt: Date
  }],
  videos: [{
    filename: String,
    originalName: String,
    path: String,
    uploadedAt: Date
  }],
  
  // ML Model Results
  mlAnalysis: {
    carbonCreditsCalculated: {
      type: Number,
      default: 0
    },
    confidenceScore: {
      type: Number,
      min: 0,
      max: 1,
      default: 0
    },
    analysisDate: {
      type: Date,
      default: Date.now
    },
    modelVersion: {
      type: String,
      default: '1.0'
    },
    rawData: {
      type: mongoose.Schema.Types.Mixed
    }
  },
  
  // Verification process
  verificationStatus: {
    type: String,
    enum: ['pending', 'under_review', 'approved', 'rejected'],
    default: 'pending'
  },
  verificationDetails: {
    verifiedBy: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Admin'
    },
    verifiedAt: Date,
    comments: String,
    approvedCarbonCredits: {
      type: Number,
      default: 0
    }
  },
  
  // Blockchain integration
  blockchainStatus: {
    type: String,
    enum: ['not_minted', 'minting', 'minted', 'failed'],
    default: 'not_minted'
  },
  transactionHash: {
    type: String,
    default: null
  },
  tokensMinted: {
    type: Number,
    default: 0
  },
  
  // Timestamps
  createdAt: {
    type: Date,
    default: Date.now
  },
  updatedAt: {
    type: Date,
    default: Date.now
  }
});

// Update the updatedAt field before saving
projectSchema.pre('save', function(next) {
  this.updatedAt = Date.now();
  next();
});

module.exports = mongoose.model('Project', projectSchema);
